modules/binary-tuple/src/main/java/org/apache/ignite/internal/binarytuple/inlineschema/TupleWithSchemaMarshalling.java

```diff
@@ -0,0 +1,378 @@
+import java.nio.ByteOrder;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.Period;
+import java.util.UUID;
+import java.util.function.BiConsumer;
+import org.apache.ignite.internal.binarytuple.BinaryTupleBuilder;
+import org.apache.ignite.internal.binarytuple.BinaryTupleReader;
+import org.apache.ignite.lang.IgniteException;
+import org.apache.ignite.marshalling.UnmarshallingException;
+import org.apache.ignite.marshalling.UnsupportedObjectTypeMarshallingException;
+import org.apache.ignite.sql.ColumnType;
+import org.apache.ignite.table.Tuple;
+import org.jetbrains.annotations.Nullable;
+
+/** Tuple with schema marshalling. */
+public final class TupleWithSchemaMarshalling {
+    private static final ByteOrder BYTE_ORDER = ByteOrder.LITTLE_ENDIAN;
+
+    /**
+     * Marshal tuple in the following format (LITTLE_ENDIAN).
+     *
+     * <pre>
+     * marshalledTuple := | size | valuePosition | binaryTupleWithSchema |
+     *
+     * size            := int32
+     * valuePosition   := int32
+     * binaryTupleWithSchema := | schemaBinaryTuple | valueBinaryTuple |
+     * schemaBinaryTuple := | column1 | ... | columnN |
+     * column              := | columnName | columnType |
+     * columnName          := string
+     * columnType          := int32
+     * valueBinaryTuple := | value1 | ... | valueN |.
+     * </pre>
+     */
+    public static byte @Nullable [] marshal(@Nullable Tuple tuple) {
+        if (tuple == null) {
+            return null;
+        }
+
+        // Allocate all the memory we need upfront.
+        int size = tuple.columnCount();
+        Object[] values = new Object[size];
+        String[] columns = new String[size];
+        ColumnType[] types = new ColumnType[size];
+
+        // Fill in the values, column names, and types.
+        for (int i = 0; i < size; i++) {
+            var value = tuple.value(i);
+            values[i] = value;
+            columns[i] = tuple.columnName(i);
+            types[i] = inferType(value);
+        }
+
+        ByteBuffer schemaBf = schemaBuilder(columns, types).build();
+        ByteBuffer valueBf = valueBuilder(columns, types, values).build();
+
+        byte[] schemaArr = getByteArray(schemaBf);
+        byte[] valueArr = getByteArray(valueBf);
+        // Size: int32 (tuple size), int32 (value offset), schema, value.
+        byte[] result = new byte[4 + 4 + schemaArr.length + valueArr.length];
+        ByteBuffer buff = ByteBuffer.wrap(result).order(BYTE_ORDER);
+
+        // Put the size of the schema in the first 4 bytes.
+        buff.putInt(0, size);
+
+        // Put the value offset in the second 4 bytes.
+        int offset = schemaArr.length + 8;
+        buff.putInt(4, offset);
+
+        System.arraycopy(schemaArr, 0, result, 8, schemaArr.length);
+        System.arraycopy(valueArr, 0, result, schemaArr.length + 8, valueArr.length);
```

Comment:
Remove `getByteArray` method, it performs unnecessary allocations. You can copy from one `ByteBuffer` to another directly.

```suggestion
        int schemaBufLen = schemaBf.remaining();
        int valueBufLen = valueBf.remaining();

        // Size: int32 (tuple size), int32 (value offset), schema, value.
        byte[] result = new byte[4 + 4 + schemaBufLen + valueBufLen];
        ByteBuffer buff = ByteBuffer.wrap(result).order(BYTE_ORDER);

        // Put the size of the schema in the first 4 bytes.
        buff.putInt(size);

        // Put the value offset in the second 4 bytes.
        int offset = schemaBufLen + 8;
        buff.putInt(offset);

        buff.put(schemaBf);
        buff.put(valueBf);
```